### PR TITLE
Implement builtin type `Signal`

### PR DESCRIPTION
**Pull request by TitanNano**
Implementation of the built-in type `Signal`.

The implementation is heavily based on the already present `Callable` type.

### CODE REVIEW

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-570

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-570

**Comment by TitanNano** [comment]
@Bromeon I'm not quite sure what is going on with the crash in 4.0

File: godot-core/src/builtin/signal.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+use crate::builtin::{inner, Array, Callable, Dictionary, StringName, Variant};
+use crate::engine::{global::Error, Object};
+use crate::obj::bounds::DynMemory;
+use crate::obj::{Bounds, Gd, GodotClass, InstanceId};
+use sys::{ffi_methods, GodotFfi};
+
+/// A `Signal` represents a signal of an Object instance in Godot.
+///
+/// Signals are composed of a reference to an `Object` and the name of the signal on this object.
+#[repr(C, align(8))]
```
Comment (Bromeon) [inline godot-core/src/builtin/signal.rs:24]: Why the `align(8)`?

File: godot-core/src/builtin/signal.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+// SAFETY:
+// The `opaque` in `Signal` is just a pair of pointers, and requires no special initialization or cleanup
+// beyond what is done in `from_opaque` and `drop`. So using `*mut Opaque` is safe.
+unsafe impl GodotFfi for Signal {
+    fn variant_type() -> sys::VariantType {
+        sys::VariantType::Signal
+    }
+
+    ffi_methods! { type sys::GDExtensionTypePtr =  *mut Opaque;
```
Comment (Bromeon) [inline godot-core/src/builtin/signal.rs:165]:
```suggestion
    ffi_methods! { type sys::GDExtensionTypePtr = *mut Opaque;
```

File: godot-core/src/builtin/signal.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /// Connects this signal to the specified callable.
+    ///
+    /// Optional flags can be also added to configure the connection's behavior (see [`ConnectFlags`](crate::engine::object::ConnectFlags) constants).
+    /// You can provide additional arguments to the connected callable by using `Callable::bind`.
+    ///
+    /// A signal can only be connected once to the same [`Callable`]. If the signal is already connected,
+    /// returns [`Error::ERR_INVALID_PARAMETER`] and
+    /// pushes an error message, unless the signal is connected with [`ConnectFlags::CONNECT_REFERENCE_COUNTED`](crate::engine::object::ConnectFlags::CONNECT_REFERENCE_COUNTED).
```
Comment (Bromeon) [inline godot-core/src/builtin/signal.rs:60]:
```suggestion
    /// pushes an error message, unless the signal is connected with [`ConnectFlags::CONNECT_REFERENCE_COUNTED`](crate::engine::object::ConnectFlags::REFERENCE_COUNTED).
```

The other one is deprecated.

File: godot-core/src/builtin/signal.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    /// Create a signal for the signal `object::signal_name`.
+    ///
+    /// _Godot equivalent: `Signal(Object object, StringName signal)`_
+    pub fn from_object_signal<T, S>(object: &Gd<T>, signal_name: S) -> Self
+    where
+        T: GodotClass,
+        S: Into<StringName>,
+    {
+        let signal = signal_name.into();
```
Comment (Bromeon) [inline godot-core/src/builtin/signal.rs:42]:
```suggestion
        let signal_name = signal_name.into();
```

File: godot-core/src/builtin/signal.rs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    ///
+    /// _Godot equivalent: `Signal()`_
+    pub fn invalid() -> Self {
+        unsafe {
+            Self::from_sys_init(|self_ptr| {
+                let ctor = sys::builtin_fn!(signal_construct_default);
+                ctor(self_ptr, ptr::null_mut())
+            })
+        }
+    }
```
Comment (Bromeon) [inline godot-core/src/builtin/signal.rs:149]: I would list constructors at the beginning of the `impl` block.

Also, should be "signal", not "object" in the doc.

**Review by Bromeon — COMMENTED** [review]
Thanks a lot!

Regarding the crash, it seems `test_varcall_signal` fails. It may need more debugging, but from a distance, it could be that the conversion between `Variant` and `Signal` is not correctly implemented for Godot 4.0. 

Maybe you could add tests in `variant_conversions()` (`variant_test.rs`) for signals -- and see if those pass or not? Feel free to use [`#[itest(focus)]`](https://godot-rust.github.io/book/contribute/dev-tools.html#integration-tests).

File: godot-core/src/builtin/signal.rs (hunk 1 above)
Comment (TitanNano) [reply godot-core/src/builtin/signal.rs:24]: Callable is also using `align(8)` and the Godot classes for both use `alignas(8)`, so I thought it was necessary.

https://github.com/godotengine/godot/blob/b7449a3d1aebd34026c7a2279a1b18ff0f02ccfa/core/variant/callable.h#L163-L171

Comment (Bromeon) [reply godot-core/src/builtin/signal.rs:24]: Makes sense!

**Review by Bromeon — APPROVED** [review]
Thanks a lot!

**Pull request merged**